nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py

```diff
@@ -160,66 +177,137 @@ def __getitem__(self, idx):
             raise e
         return self._process_example(example)
 
+    def _process_prompt(self, contexts: List[str], label: str):
+        """
+        Combine contexts and label into a list of strings and a list of keys based on template.
+        """
+        prompt_strings = []
+        prompt_keys = []
+
+        # split with template
+        if self.prompt_template is not None:
+            # context placeholder
+            context_phs = [f'{{{ck}}}' for ck in self.context_key]
+            for cph in context_phs:
+                assert cph in self.prompt_template, f'{cph} must in {self.prompt_template}'
+            context_ph_to_context_s = {cph: cs for cph, cs in zip(context_phs, contexts)}
+            context_ph_to_context_k = {cph: ck for cph, ck in zip(context_phs, self.context_key)}
+
+            # label placeholder
+            label_ph = f'{{{self.label_key}}}'
+            assert label_ph in self.prompt_template, f'{label_ph} must in {self.prompt_template}'
+            assert self.prompt_template.index(label_ph) == len(self.prompt_template) - len(
+                label_ph
+            ), f'{{{self.label_key}}} must be at the end of prompt_template.'
+
+            # positions
+            context_positions = [
+                p
+                for cph in context_phs
+                for p in (self.prompt_template.index(cph), self.prompt_template.index(cph) + len(cph))
+            ]
+            context_positions = [0] + context_positions + [self.prompt_template.index(label_ph)]
+            context_positions = sorted(context_positions)
+
+            # iterate each position bucket
+            for i in range(len(context_positions) - 1):
+                pi = context_positions[i]
+                pj = context_positions[i + 1]
+                # prev_s end space + curr_s start space
+                leading_space = (
+                    0
+                    if i == 0
+                    else int(self.prompt_template[pi - 1] == ' ')
+                    + len(self.prompt_template[pi:pj])
+                    - len(self.prompt_template[pi:pj].lstrip(' '))
+                )
+                cs = context_ph_to_context_s.get(self.prompt_template[pi:pj], self.prompt_template[pi:pj]).strip(' ')
+                if len(cs) > 0:
+                    prompt_strings.append(' ' * leading_space + cs if self.tokenizer_space_sensitive else cs)
+                    prompt_keys.append(context_ph_to_context_k.get(self.prompt_template[pi:pj], '<template>'))
+
+            leading_space = int(self.prompt_template[pj - 1] == ' ')
+            prompt_strings.append(' ' * leading_space + label if self.tokenizer_space_sensitive else label)
+            prompt_keys.append(self.label_key)
```

Comment:
@hsiehjackson might be best if this was in a separate method with a doc string explaining what is happening.

Reply:
@arendu I add more comments for this logic. Since we remove `separate_prompt_and_response_with_newline`, I think I can directly add doc string in this function